src/util/result.h

```diff
@@ -8,16 +8,133 @@
 #include <attributes.h>
 #include <util/translation.h>
 
+#include <memory>
+#include <optional>
+#include <tuple>
 #include <variant>
+#include <vector>
 
 namespace util {
+namespace detail {
+//! Empty string list
+const std::vector<bilingual_str> EMPTY_LIST{};
 
+//! Helper to move elements from one container to another.
+template <typename T>
+void MoveElements(T& src, T& dest)
+{
+    dest.insert(dest.end(), std::make_move_iterator(src.begin()), std::make_move_iterator(src.end()));
+    src.clear();
+}
+
+//! Error information only allocated if there are errors or warnings.
+template <typename F>
+struct ErrorInfo {
+    std::optional<std::conditional_t<std::is_same<F, void>::value, std::monostate, F>> failure;
+    std::vector<bilingual_str> errors;
+    std::vector<bilingual_str> warnings;
+};
+
+//! Result base class which is inherited by Result<T, F>.
+template <typename T, typename F>
+class ResultBase;
+
+//! Result base specialization for empty (T=void) value type. Holds error
+//! information and provides accessor methods.
+template <typename F>
+class ResultBase<void, F>
+{
+protected:
+    std::unique_ptr<ErrorInfo<F>> m_info;
+
+    template <typename... Args>
+    void InitFailure(Args&&... args)
+    {
+        if (!m_info) m_info = std::make_unique<ErrorInfo<F>>();
+        m_info->failure.emplace(std::forward<Args>(args)...);
+    }
+    void InitValue() {}
+    template <typename O>
+    void MoveValue(O& other) {}
+    void DestroyValue() {}
+
+public:
+    void AddError(bilingual_str error)
+    {
+        if (error.empty()) return;
+        if (!m_info) m_info = std::make_unique<ErrorInfo<F>>();
+        m_info->errors.emplace_back(std::move(error));
+    }
+
+    void AddWarning(bilingual_str warning)
+    {
+        if (warning.empty()) return;
+        if (!m_info) m_info = std::make_unique<ErrorInfo<F>>();
+        m_info->warnings.emplace_back(std::move(warning));
+    }
+
+    //! Success check.
+    operator bool() const { return !m_info || !m_info->failure; }
+
+    //! Error retrieval.
+    template <typename _F = F>
+    std::enable_if_t<!std::is_same<_F, void>::value, const _F&> GetFailure() const { assert(!*this); return *m_info->failure; }
+    const std::vector<bilingual_str>& GetErrors() const { return m_info ? m_info->errors : EMPTY_LIST; }
+    const std::vector<bilingual_str>& GetWarnings() const { return m_info ? m_info->warnings : EMPTY_LIST; }
```

Comment:
Could probably make sense to split this up and use it there: https://github.com/bitcoin/bitcoin/pull/25616#discussion_r942704207

Reply:
No, I was wondering if `GetWarnings` could be added by itself in the other pull, maybe conflicting with this one, but otherwise unrelated from this one.

Reply:
This would allow reviewers to just look at the `GetWarnings` implementation, API, and proposed usage, without having to think about `union/void/...` etc.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25665#discussion_r945881360

> This would allow reviewers to just look at the `GetWarnings` implementation, API, and proposed usage, without having to think about `union/void/...` etc.

Oh. I guess there could be a third PR independent of this PR and #25722 that adds a `std::vector<bilingual_str> m_warnings;` field to `util::Result` and `AddWarning()/GetWarnings()` methods. I don't think it would be as useful as you might be thinking without template constructors that can move errors & warnings from one result object to another, though.

Maybe if you are interested you could open a PR that does this? I'd be happy to review it. I'm just not sure exactly what you want to do here because the goal of #25722 is a lot broader. Not just dropping `std::vector<bilingual_str>& warnings` output arguments but also dropping `DatabaseStatus& status` output arguments and returning errors and warnings directly from nested calls.

Reply:
The thing is that you are basically adding a ton of dead code in this pull. `bitcoind` compiles fine without it, so I might be picky, but my preference would be to only add the code when it is needed. This also makes it easier for reviewers because they can think about actual use cases and don't have to imagine them from the unit tests.


```diff
diff --git a/src/util/result.cpp b/src/util/result.cpp
index 9526b5b785..199a6579d7 100644
--- a/src/util/result.cpp
+++ b/src/util/result.cpp
@@ -3,26 +3,8 @@
 // file COPYING or https://www.opensource.org/licenses/mit-license.php.
 
 #include <util/result.h>
-#include <util/string.h>
 
 namespace util {
 namespace detail {
-bilingual_str JoinMessages(const std::vector<bilingual_str>& errors, const std::vector<bilingual_str> warnings)
-{
-    bilingual_str result;
-    for (const auto& messages : {errors, warnings}) {
-        for (const auto& message : messages) {
-            if (!result.empty()) result += Untranslated(" ");
-            result += message;
-        }
-    }
-    return result;
-}
-
-void MoveMessages(std::vector<bilingual_str>& src, std::vector<bilingual_str>& dest)
-{
-    dest.insert(dest.end(), std::make_move_iterator(src.begin()), std::make_move_iterator(src.end()));
-    src.clear();
-}
 } // namespace detail
 } // namespace util
diff --git a/src/util/result.h b/src/util/result.h
index 32fe40763f..8eb5b552d3 100644
--- a/src/util/result.h
+++ b/src/util/result.h
@@ -17,21 +17,11 @@
 
 namespace util {
 namespace detail {
-//! Empty string list
-const std::vector<bilingual_str> EMPTY_LIST{};
-
-//! Helper function to join messages in space separated string.
-bilingual_str JoinMessages(const std::vector<bilingual_str>& errors, const std::vector<bilingual_str> warnings);
-
-//! Helper function to move messages from one vector to another.
-void MoveMessages(std::vector<bilingual_str>& src, std::vector<bilingual_str>& dest);
-
-//! Error information only allocated if there are errors or warnings.
+//! Error information only allocated if there is an error.
 template <typename F>
 struct ErrorInfo {
     std::optional<std::conditional_t<std::is_same<F, void>::value, std::monostate, F>> failure{};
-    std::vector<bilingual_str> errors;
-    std::vector<bilingual_str> warnings;
+    bilingual_str error;
 };
 
 //! Result base class which is inherited by Result<T, F>.
@@ -66,25 +56,7 @@ protected:
 public:
     void AddError(bilingual_str error)
     {
-        if (!error.empty()) Info().errors.emplace_back(std::move(error));
-    }
-
-    void AddWarning(bilingual_str warning)
-    {
-        if (!warning.empty()) Info().warnings.emplace_back(std::move(warning));
-    }
-
-    //! Operator moving warning and error messages from other result to this.
-    //! Only moves message strings, does not change success or failure values of
-    //! either Result object.
-    template<typename O>
-    O&& operator<<(O&& other)
-    {
-        if (other.m_info) {
-            if (!other.m_info->errors.empty()) MoveMessages(other.m_info->errors, Info().errors);
-            if (!other.m_info->warnings.empty()) MoveMessages(other.m_info->warnings, Info().warnings);
-        }
-        return std::forward<O>(other);
+        Info().error = std::move(error);
     }
 
     //! Success check.
@@ -93,8 +65,7 @@ public:
     //! Error retrieval.
     template <typename _F = F>
     std::enable_if_t<!std::is_same<_F, void>::value, const _F&> GetFailure() const { assert(!*this); return *m_info->failure; }
-    const std::vector<bilingual_str>& GetErrors() const { return m_info ? m_info->errors : EMPTY_LIST; }
-    const std::vector<bilingual_str>& GetWarnings() const { return m_info ? m_info->warnings : EMPTY_LIST; }
+    bilingual_str GetError() const { return m_info ? m_info->error : bilingual_str{}; }
 };
 
 //! Result base class for T value type. Holds value and provides accessor methods.
@@ -145,9 +116,6 @@ public:
 struct Error {
     bilingual_str message;
 };
-struct Warning {
-    bilingual_str message;
-};
 
 //! The util::Result class provides a standard way for functions to return error
 //! and warning strings in addition to optional result values.
@@ -190,23 +158,11 @@ protected:
         }, std::forward<Args>(args)...);
     }
 
-    template <typename Fn, typename... Args>
-    void Construct(const Fn& fn, Warning warning, Args&&... args)
-    {
-        this->AddWarning(std::move(warning.message));
-        Construct(fn, std::forward<Args>(args)...);
-    }
-
-    template <typename Fn, typename OT, typename OF, typename... Args>
-    void Construct(const Fn& fn, Result<OT, OF>&& other, Args&&... args)
-    {
-        *this << other;
-        Construct(fn, std::forward<Args>(args)...);
-    }
-
     void MoveConstruct(Result& other)
     {
-        *this << other;
+        if (other.m_info) {
+            this->AddError(std::move(other.m_info->error));
+        }
         if (other) this->MoveValue(other); else this->Info().failure = std::move(other.m_info->failure);
     }
 
@@ -233,7 +189,7 @@ public:
 //! are present. More complicated applications should use GetErrors() and
 //! GetWarning() methods directly.
 template <typename T, typename F>
-bilingual_str ErrorString(const Result<T, F>& result) { return detail::JoinMessages(result.GetErrors(), result.GetWarnings()); }
+bilingual_str ErrorString(const Result<T, F>& result) { return result.GetError(); }
 } // namespace util
 
 #endif // BITCOIN_UTIL_RESULT_H

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25665#discussion_r957494674

> The thing is that you are basically adding a ton of dead code in this pull

Interesting and thanks for the diff! Now I understand what you are suggesting, and I think I'd be ok with making that change. I wasn't really looking at that code as dead because I'm immediately using it in the next pull #25722 and also using it in unit tests this PR. I also didn't consider it to be a ton of code, since the code in your diff is just getter/setter functions that provide access to errors/warnings variables. But no objection to splitting it off into another pull.

Reply:
Yeah, I don't feel too strong as well. Though, if you keep it, it would be good to fixup the typo: https://github.com/bitcoin/bitcoin/pull/25665#discussion_r957495310

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25665#discussion_r958050929

> Yeah, I don't feel too strong as well. Though, if you keep it, it would be good to fixup the typo: [#25665 (comment)](https://github.com/bitcoin/bitcoin/pull/25665#discussion_r957495310)

Ok, I'm hedging right now by moving multiple error & warning support into a separate commit but not a separate PR. I maybe have a slight preference for just squashing the commits again to reduce churn, but I'm also fine with keeping separate commits, or moving the commit to another PR if one of those alternatives seems better, so let me know!

Also fixed the typo